File: typescript/Style/render.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import * as INDEX from "../data/index.js";
import * as CACHE from "../data/cache.js";
import * as LOADPREFIX from "./prefix.js";
// import HASHRULE from "../hashrule.js";
import * as _Style from "../type/style.js";

import Use from "../utils/main.js";

type t_styleSorceTemplate = Record<string, string | object>;

function objectSwitch(srcObject: Record<string, any>): Record<string, any> {
	if (!srcObject || typeof srcObject !== "object") {
		return {};
	}

	const output: Record<string, any> = {};

	for (const outerKey in srcObject) {
		const innerObject = srcObject[outerKey];
		if (typeof innerObject === "object" && innerObject !== null) {
			for (const innerKey in innerObject) {
				if (!output[innerKey]) { output[innerKey] = {}; }
				output[innerKey][(innerKey[0] === '@' || innerKey === "") ? outerKey : `& ${outerKey}`] = innerObject[innerKey];
				// if(outerKey === ""){
				// 	output[innerKey][(innerKey[0] === '@' || innerKey === "") ? outerKey : `& ${outerKey}`] = innerObject[innerKey];
				// }else{
				// 	HASHRULE.WRAPPER(output[innerKey], JSON.parse(outerKey), innerObject[innerKey]);
				// }
			}
		}
	}

	return output;
}

function styleSwitch(object: Record<string, Record<string, object | string>>) {
	const result: Record<string, object> = {};
	const inits: string[] = [], mins: string[] = [], maxs: string[] = [], flats: string[] = [];

	const switched = objectSwitch(object);
	Object.keys(switched).forEach((key) => {
		const min = key.indexOf("min");
		const max = key.indexOf("max");
		if (key !== "") {
			if (min === -1 && max === -1) { inits.push(key); }
			else if (min < max) { mins.push(key); }
			else if (min > max) { maxs.push(key); }
			else if (min === max) { flats.push(key); }
		}
	});

	inits.forEach(key => result[key] = switched[key]);
	Object.assign(result, switched[""]);
	[...flats.sort(), ...mins.sort().reverse(), ...maxs.sort()].forEach((key) => (result[key] = switched[key]));
	return result;
}

function LoadVendors(collection = {}, vendor = "") {
	return vendor == ""
		? CACHE.ROOT.vendors.filter((ven) =>
			!Object.prototype.hasOwnProperty.call(collection, ven),
		) : [vendor];
}

function partialsArrayPrefixer(object: t_styleSorceTemplate, vendors = LoadVendors()) {
	const result: [string, string | object][] = [];
	Object.entries(object).forEach(([key, value]) => {
		if (typeof value === "object") {
			if (Object.keys(value).length) {
				result.push([key, value]);
			}
		} else if (key[0] === "@") {
			Object.values(LOADPREFIX.forAtRule(key, vendors)).forEach((r) =>
				result.push([r + ";", ""]),
			);
		} else {
			LOADPREFIX.LoadProps(key, value, vendors).forEach(([k, v]) => {
				if (k === key || !object[k]) { result.push([k + ":", v + ";"]); }
			});
		}
	});
	return result;
}

// Pending to handle states &:* states.

function unNester(selector = "", object: object = {}, cumulates: object = {}) {
	const
		compounds: t_styleSorceTemplate = {},
		pseudoclass: t_styleSorceTemplate = {},
		pseudoelement: t_styleSorceTemplate = {},
		children: t_styleSorceTemplate = {},
		myself: t_styleSorceTemplate = {};
	const holder: t_styleSorceTemplate = myself[selector] = {};

	Object.entries(object as t_styleSorceTemplate).forEach(([subSelector, subContent]) => {
		if (typeof subContent === "object") {
			if (subSelector[0] === "&") {
				const xelector = selector + subSelector.slice(1);
				if (subSelector[1] === ":") {
					unNester(xelector, subContent, subSelector[2] === ":" ? pseudoelement : pseudoclass);
				} else if (subSelector[1] === " ") {
					unNester(xelector, subContent, children);
				} else {
					unNester(xelector, subContent, compounds);
				}
			} else {
				unNester(subSelector, subContent, holder);
			}
		} else {
			holder[subSelector] = subContent;
		}
	});

	Object.assign(cumulates, compounds, pseudoclass, myself, pseudoelement, children);
	return cumulates as t_styleSorceTemplate;
}

function _objectCompose(
	object: t_styleSorceTemplate,
	minify = false,
	vendors = LoadVendors(),
	first = true,
) {
	const tab = minify ? "" : "  ",
		space = minify ? "" : " ",
		styleSheet: string[] = [];

	partialsArrayPrefixer(object, vendors).forEach(([key, value]) => {
		if (typeof value === "object") {
			if (Object.keys(value).length) {
				if (!minify && first) { styleSheet.push(""); }
				if (key[0] === "@") {
					const atPrefixes = LOADPREFIX.forAtRule;
					Object.entries(atPrefixes(key, vendors)).forEach(
						([vendor, selector]) => {
							const composedObject = _objectCompose(
								value as t_styleSorceTemplate,
								minify,
								LoadVendors(atPrefixes, vendor),
								false,
							);
							if (composedObject.length) {
								styleSheet.push(
									selector,
									"{",
									...composedObject.map((i) => tab + i),
									"}",
								);
							}
						},
					);
				} else {
					const composedObject = _objectCompose(
						value as t_styleSorceTemplate,
						minify,
						vendors,
						false);
					if (Object.keys(composedObject).length) {
						styleSheet.push(...LOADPREFIX.forPseudos(key, vendors));
						styleSheet.push(
							"{",
							...composedObject.map((i) => tab + i),
							"}",
						);
					}
				}
			}
		} else if (key[0] === "@") {
			styleSheet.push(key);
		} else {
			styleSheet.push(key + space + value);
		}
	});

	return styleSheet;
}




function ComposePrefixed(array: [string, string | object][], minify = !CACHE.STATIC.DEBUG) {
	const styleSheet: string[] = [];

	array.forEach(([key, value]) => {
		if (typeof value === "object") {
			const unNested = unNester(key, value);
			if (Object.keys(unNested).length) {
				styleSheet.push(..._objectCompose(unNested, minify));
			}
		} else {
			styleSheet.push(..._objectCompose({ [key]: value }, minify));
		}
	});

	return styleSheet.join(minify ? "" : "\n");
}


function ComposeSwitched(selectorIndexObject: Record<string, number>, minify = !CACHE.STATIC.DEBUG) {
	const object = styleSwitch(
		Object.entries(selectorIndexObject).reduce((A: Record<string, Record<string, object>>, [selector, index]) => {
			A[selector] = INDEX.FETCH(index).style_object;
			return A;
		}, {}),
	);

	return ComposePrefixed(Object.entries(object), minify);
}


function ArtifactPartial(object: object, minify = true) {
	const array = Object.entries(object);
	const styleSheet: string[] = [];
	const tab = minify ? "" : "  ";

	array.forEach(([key, value]) => {
		if (typeof value === "object") {
			if (Object.keys(value).length) {
				styleSheet.push(
					key,
					"{",
					...ArtifactPartial(value).map((i) => tab + i),
					"}",
				);
			}
		} else if (key[0] === "@") {
			styleSheet.push(key + ";");
		} else {
			styleSheet.push(key + ": " + value + ";");
		}
	});

	return styleSheet;
}

function ComposeArtifact(index: number): _Style.ExportStyle {
	const style = INDEX.FETCH(index);

	let element = "";
	if (style.snippet_staple.length) {
		element = "staple";
	} else if (style.metadata.summon.length) {
		element = "summon";
	} else {
		element = "style";
	};

	const symclass = style.symclass.includes("$$")
		? (style.symclass.startsWith("$") ? `-${style.symclass}` : style.symclass)
		: `---${Use.string.enCounter(style.index || 0)}`;

	const stylesheet = Object.fromEntries(Object.entries(style.style_object).map(([k, v]) => {
		return [k, ArtifactPartial(v).join("")];
	}));

	const innertext = style.snippet_staple || style.metadata.summon || ArtifactPartial(style.snippet_style).join("");
	
	return {
		element,
		symclass,
		innertext,
		stylesheet,
		attachments: []
	};
}

export default {
	Prefixed: ComposePrefixed,
	Switched: ComposeSwitched,
	Artifact: ComposeArtifact,
};

File: gosource/compose/render.go
package compose

import (
	_strings_ "strings"
)

// func render_LoadVendors(collection map[string]string, vendor string) []string {
// 	result := []string{}
// 	if vendor == "" {
// 		for _, ven := range vendor_Providers {
// 			if _, stat := collection[ven]; stat {
// 				result = append(result, ven)
// 			}
// 		}
// 	} else {
// 		result = append(result, vendor)
// 	}
// 	return result
// }

func render_PartialsArrayPrefixer(object map[string]any, vendors []string) [][2]any {
	var result [][2]any

	for key, value := range object {
		switch val := value.(type) {
		case map[string]any:
			if len(val) > 0 {
				result = append(result, [2]any{key, val})
			}
		case string:
			if _strings_.HasPrefix(key, "@") {
				for _, r := range prefix_ForAtRule(key, vendors) {
					result = append(result, [2]any{r + ";", ""})
				}
			} else if valstr, ok := value.(string); ok {
				for _, kv := range prefix_LoadProps(key, valstr, vendors) {
					k, v := kv[0], kv[1]
					if k == key || object[k] == nil {
						result = append(result, [2]any{k + ":" + v + ";"})
					}
				}
			}
		}
	}
	return result
}

// Pending to handle states &:* states.

func render_UnNester(selector string, value map[string]any, cumulates map[string]any) map[string]any {
	compounds := map[string]any{}
	pseudoclass := map[string]any{}
	pseudoelement := map[string]any{}
	children := map[string]any{}
	myself := map[string]any{}
	holder := map[string]any{}

	for key, val := range value {
		switch val_typed := val.(type) {
		case map[string]any:
			if key[0] == '&' {
				xelector := selector + key[1:]
				if key[1] == ':' {
					if key[2] == ':' {
						render_UnNester(xelector, val_typed, pseudoelement)
					} else {
						render_UnNester(xelector, val_typed, pseudoclass)
					}
				} else if key[1] == ' ' {
					render_UnNester(xelector, val_typed, children)
				} else {
					render_UnNester(xelector, val_typed, compounds)
				}
			} else {
				render_UnNester(key, val_typed, holder)
			}
		default:
			holder[key] = val
		}
	}

	for _, m := range []map[string]any{compounds, pseudoclass, myself, pseudoelement, children, holder} {
		for k, v := range m {
			cumulates[k] = v
		}
	}
	return cumulates
}

func render_ObjectCompose(object map[string]any, minify bool, vendors []string, first bool) []string {
	stylesheet := []string{}
	tab := "  "
	space := " "
	if minify {
		tab = ""
		space = ""
	}

	for _, kv := range render_PartialsArrayPrefixer(object, vendors) {
		key_typed, key_ok := kv[0].(string)
		val := kv[1]
		if key_ok {
			switch val_typed := val.(type) {
			case map[string]any:
				if len(val_typed) > 0 {
					if !minify && first {
						stylesheet = append(stylesheet, "")
					}
					if key_typed[0] == '@' {
						for vendor, selector := range prefix_ForAtRule(key_typed, vendors) {
							composed := render_ObjectCompose(val_typed, minify, []string{vendor}, false)
							if len(composed) > 0 {
								stylesheet = append(stylesheet, selector)
								stylesheet = append(stylesheet, "{")
								for _, i := range composed {
									stylesheet = append(stylesheet, tab+i)
								}
								stylesheet = append(stylesheet, "}")
									
							}

						}
					} else {
						composed := render_ObjectCompose(val_typed, minify, vendors, false)
						if !minify {
							for index, line := range composed {
								composed[index] = tab + line
							}
						}
						if len(composed) > 0 {
							stylesheet = append(stylesheet, prefix_ForPseudos(key_typed, vendors)...)
							stylesheet = append(stylesheet, "{")
							stylesheet = append(stylesheet, composed...)
							stylesheet = append(stylesheet, "}")
						}
					}
				}
			case string:
				if key_typed[0] == '@' {
					stylesheet = append(stylesheet, key_typed)
				} else {
					stylesheet = append(stylesheet, key_typed+space+val_typed)
				}
			}
		}
	}

	return stylesheet
}

func Render(array [][2]any, minify bool)string {
	stylesheet := []string{}
	breaks := "\n"
	if minify {
		breaks = ""
	}

	for _, kv := range array {
		key := kv[0]
		key_typed, key_ok := key.(string)

		if key_ok {
			val := kv[1]
			flattened := map[string]any{} 

			switch val_typed := val.(type) {
			case map[string]any: 
				flattened = render_UnNester(key_typed, val_typed, map[string]any{})
			case string:
				flattened = map[string]any{ key_typed: val_typed }
			}

			composed := render_ObjectCompose(flattened, minify, vendor_Providers, true)
			stylesheet = append(stylesheet, composed...)
		}
	}

	return _strings_.Join(stylesheet, breaks)
}
